src/medicine/main.py

```diff
@@ -1,47 +1,57 @@
-import pickle
-from icecream import ic
-from medicine.medicine_classes import Medicine
+import json
+from typing import List
+from medicine.medicine import Medicine
 
 
-def save_dict_to_file(dictionary, file_name):
-    assert isinstance(dictionary, dict), 'dictionary is supposed to be DICT'
-    assert isinstance(file_name, str), 'file_name is supposed to be STR'
-    import pickle  # służy do zapisywania binarnego plików
-    with open(file_name, 'wb') as file:  # tryb wb też
-        pickle.dump(dictionary, file)  # żeby słownik zapisywać do pliku
+def export_to_file(filename: str, items: List[Medicine]):
+    with open(filename, 'w') as f:
+        json.dump([item.serialize() for item in items], f, indent=4)
+
+
+def import_from_file(filename: str) -> List[Medicine]:
+    with open(filename) as f:
+        return [Medicine(**item) for item in json.load(f)]
 
 
 def how_much_in_mg(*medicines):
     """Takes medicine's names (STR) and returns amount of the medicine in milligrams"""
+    # Ja wolę taki zapis, ale obie działają równie dobrze
+    # return sum([medicine.current_mg for medicine in medicines])
     sum_of_mg = []  # to this list I add milligrams multiplied by the number of pills
     for medicine in medicines:
-        assert isinstance(medicine, Medicine), 'medicines is supposed to be medicine_classes.Medicine'
-        amount_of_pills = medicine.current_amount
-        mg_per_pill = medicine.dose
-        amount_of_mg = amount_of_pills * mg_per_pill
-        sum_of_mg.append(amount_of_mg)
+        sum_of_mg.append(medicine.current_mg)
     return sum(sum_of_mg)
 
-def read_pickle_file(file_name):
-    with open(file_name, 'rb') as file:
-        data = pickle.load(file)
-    return data
 
 vetira_500 = Medicine('vetira', 500, 1500, 10)
 vetira_750 = Medicine('vetira', 750, 1500, 10)
 neurotop_600 = Medicine('neurotop', 600, 1200, 40)
 lacosamide_100 = Medicine('lacosamide', 100, 200, 50)
 
 
-# ic(Medicine.instances)
-
-
 def run():
-    data = read_pickle_file('instances.p')
-    neurotop_600.remove_instance_from_file()
-    ic(read_pickle_file('instances.p'))
-
-    print(data['lacosamide_100'][1])
+    data = import_from_file('instances.json')
+
+    # Tu różne operacje które dodają/usuwają/zmieniają stan różnych leków.
+    # Można za każdym razem zapisać do pliku, albo tylko raz na końcu. Obie
+    # podejścia mają swoje wady i zalety:
+    #
+    # Zapis za każdym razem jest wolniejszy (aczkolwiek w tym przypadku to nie istotne), ale jak np.
+    # byś to robił tak że ci wypisuje co masz brać lek po leku, i przy każdym leku zatwierdzasz że wziąłeś,
+    # to dobrze by było po każdym zatwierdzeniu zapisać.
+    #
+    # Jeżeli to ma działać bardziej na zasadzie że wypisujesz całą listę, a na końcu potwierdzasz że wszystko
+    # wziąłeś, to zapisz na końcu.
+    #
```

Comment:
Ok, ogólnie:

- jeszcze program nie jest w połowie gotowy nawet, więc pewnie wiele rzeczy może się wykrzaczać
- plik picklowy /JSONowy jest po to żeby w nim przechowywać jakiego leku kiedy ile było. Jak zacznę ogarniać sobie `datetime` albo inną bibliotekę to wtedy to w praktyce się przyda. Plan jest taki, że zapis do tego pliku będzie *tylko* wtedy, gdy będę tworzył instancję (nowy lek) lub dodawał więcej tabletek (funkcja `add_medicine`, którą muszę jeszcze porządnie rozbudować). Na razie sobie ten plik modyfikuję na bieżąco jako forma testów. 

Swoją drogą tutaj uznałem za bardzo uciążliwe tworzenie testów, gdy test musi być wykonany w środku klasy, albo gdy musi być wykonany na pliku, na którym pracuję. Więcej to czasu zabiera niż pisanie kodu (nie mówię, że nie powinienem robić testów).

Reply:
Przeczytam resztę za kilka godzin, bo muszę spadać

Reply:
spoko, więc w zasadzie jest to pewnego rodzaju dziennik operacji. Tutaj proponowałbym to prowadzić w sposób taki że każda linia jest nowym wydarzeniem, i każda linija ma datę, nazwa leku, ile przybyło (wtedy połykanie to ujemne przybycie) oraz ile myślisz że teraz zostało. Czyli taki [rejestr rachunkowy](https://pl.wikipedia.org/wiki/Regu%C5%82a_podw%C3%B3jnego_zapisu). Wtedy łatwo ci będzie odwtorzyć całą historię w razie problemów. No i też znacznie ułatwia zapis do pliku, bo to tylko
```
with open('log', 'a') as f:
    f.write(nowe_wydarzenie + '\n')
```

Testy są ogólnie bardzo uciążliwe. Ale bardzo przydatne :/ Mój nacisk na małe funkcje, najlepiej poza klasą w dużej mierze wypływa z frustracji testowaniem. Bo wtedy nie musisz tego wszystkiego robić. Jak masz jedną funkcje ktora tylko zapisuje do pliku, to względnie łatwo to przetestować, ale co ważniejsze, nie musisz się tym potem przejmować jak testujesz właściwe działania. np.
```
def write_to_file(filename, data):
    bla bla bal

def change_medicine(medicine):
    bla bla bla
    return medicine

def do_stuff(medicine):
    write_to_file(change_medicine(medicine))
```
się o wiele ławiej testuje niż
```
def do_stuff(medicine):
   bla bla bla jakieś zmiany
   zapisz zmiany do pliku
```

bo porządnie otestujesz `write_to_file` i `change_medicine` względnie łatwo, a `do_stuff` nie musisz już testować, bo nie bardzo ma jak się zepsuć. Tzn. ma, i warto choć 1 test mieć na wszelki wypadek, ale jest to o wiele łatwiej

Reply:
przeczytaj kiedy chcesz, jak będziesz miał czas. Ja już swoje napisałem, także teraz to twoja sprawa co z tym zrobisz :D To jest piękno komunikacji asynchronicznej, że nie trzeba odrazu reagować. Możesz za 3 miesiące jak byś wcześniej nie miał czasu - mi to nie przeszkodzi

Reply:
No ogólnie to staram się dekomponować problemy na wiele funkcji.

Co do sposobu zapisu - na razie mój pomysł był taki: w pliku jest słownik którego kluczem jest nazwa leku, a wartością krotka (`ilość_leku_w_mg`, `data dodania/modyfikacji`). Wtedy by wystarczyła funkcja licząca na ile dni mi starczy leku, bo ma informację ile jest leku dnia jakiegoś tam i wie jaki dzisiaj jest dzień (no i wie ile dziennie spożywam). Nie byłoby to prostrze niż dodawanie i odejmowanie w nieskończoność leku?

Plus pierwsza wersja ma być do odpalenia ręcznie raz na jakiś czas, nie będzie codziennie uruchamiana.

Reply:
No dobra, mógłbym bardziej dekomponować, ale tragedii chyba nie ma :p